apps/frontend/src/pages/Charities.tsx

```diff
@@ -0,0 +1,258 @@
+        setLoading(false);
+      }
+    };
+    fetchData();
+  }, []);
+
+  const currentCharity = Array.isArray(charities)
+    ? charities.find((charity) => new Date(charity.endDate) >= currentDate)
+    : undefined;
+
+  const pastCharities = Array.isArray(charities)
+    ? charities.filter((charity) => new Date(charity.endDate) < currentDate)
+    : [];
+
+  useEffect(() => {
+    if (currentCharity) {
+      const endTime = new Date(currentCharity.endDate).getTime();
+      const interval = setInterval(() => {
+        const now = new Date().getTime();
+        const distance = endTime - now;
+
+        if (distance < 0) {
+          clearInterval(interval);
+          setTimeRemaining("This charity event has ended.");
+        } else {
+          const days = Math.floor(distance / (1000 * 60 * 60 * 24));
+          const hours = Math.floor(
+            (distance % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60),
+          );
+          const minutes = Math.floor(
+            (distance % (1000 * 60 * 60)) / (1000 * 60),
+          );
+          const seconds = Math.floor((distance % (1000 * 60)) / 1000);
+          setTimeRemaining(`${days}d ${hours}h ${minutes}m ${seconds}s`);
+        }
+      }, 1000);
+
+      return () => clearInterval(interval);
+    }
+  }, [currentCharity]);
+
+  const handleToggleChange = (event: ChangeEvent<HTMLInputElement>) => {
+    setShowPastEvents(event.target.checked);
+  };
+
+  if (loading) {
+    return <Spinner text="Loading charities..." />;
+  }
+
+  return (
+    <>
+      <SearchBar />
+      <FormControlLabel
+        control={
+          <Switch
+            checked={showPastEvents}
+            onChange={handleToggleChange}
+            color="primary"
+            id="charitiesSwitch"
+          />
+        }
+        label={
+          showPastEvents
+            ? "Show Current Charity Event"
+            : "Show Past Charity Events"
+        }
+        sx={{ marginLeft: "20px", marginTop: "20px" }}
+      />
+      {showPastEvents ? (
+        pastCharities.length > 0 ? (
+          <>
+            <Typography
+              variant="h5"
+              component="h3"
+              gutterBottom
+              sx={{ paddingLeft: "20px", paddingTop: "20px" }}
+            >
+              Past Charity Events
+            </Typography>
+            <Grid container spacing={2} justifyContent="center">
+              {pastCharities.map((charity, index) => (
+                <Grid item xs={12} sm={6} md={4} key={index}>
+                  <CharityCard {...charity} />
+                </Grid>
+              ))}
+            </Grid>
+          </>
+        ) : (
+          <Typography
+            variant="h6"
+            component="h4"
+            gutterBottom
+            sx={{ paddingLeft: "20px", paddingTop: "20px" }}
+          >
+            No Past Charities
+          </Typography>
+        )
+      ) : (
+        <>
+          {currentCharity ? (
+            <>
```

Comment:
This fragment is unnecessary. Also it might be nice to factor one or two of these elements out in to componenets to avoid 250+ line files